core/bootstrap/src/main/java/org/eclipse/dataspaceconnector/system/runtime/BaseRuntime.java

```diff
@@ -0,0 +1,183 @@
+    }
+
+    /**
+     * Create a list of {@link ServiceExtension}s. By default this is done using the ServiceLoader mechanism. Override if
+     * e.g. a custom DI mechanism should be used.
+     *
+     * @param context A context to which all the service extensions should be registered
+     * @return a list of {@code ServiceExtension}s
+     */
+    @NotNull
+    protected List<ServiceExtension> createExtensions(ServiceExtensionContext context) {
+        return context.loadServiceExtensions();
+    }
+
+    /**
+     * Create a {@link ServiceExtensionContext} that will be used in this runtime. If e.g. a custom DI mechanism should be used,
+     * this would likely need to be overridden.
+     *
+     * @param typeManager The TypeManager (for JSON de-/serialization)
+     * @param monitor     a Monitor
+     * @return a {@code ServiceExtensionContext}
+     */
+    @NotNull
+    protected ServiceExtensionContext createContext(TypeManager typeManager, Monitor monitor) {
+        return new DefaultServiceExtensionContext(typeManager, monitor);
+    }
+
+    /**
+     * Hook that is called when a runtime is shutdown (e.g. after a CTRL-C command on a command line). It is highly advisable to
+     * forward this signal to all extensions through their {@link ServiceExtension#shutdown()} callback.
+     *
+     * @param serviceExtensions All extensions that should receive the shutdown signal.
+     * @param monitor           A monitor - should you need one.
+     */
+    protected void shutdown(List<ServiceExtension> serviceExtensions, Monitor monitor) {
+        ListIterator<ServiceExtension> iter = serviceExtensions.listIterator(serviceExtensions.size());
+        while (iter.hasPrevious()) {
+            iter.previous().shutdown();
+        }
+        monitor.info("Connector shutdown complete");
+    }
+
+    /**
+     * Hook point to initialize the vault. By default this calls {@link org.eclipse.dataspaceconnector.system.ExtensionLoader#loadVault(ServiceExtensionContext)}
+     * to instantiate and configure the vault.
+     * <p>
+     * <strong>Do not use this to provide a custom Vault implementation! Use the extension mechanism for that! </strong>
+     *
+     * @param context An {@code ServiceExtensionContext} to resolve the {@code Vault} from.
+     */
+    protected void initializeVault(ServiceExtensionContext context) {
+        loadVault(context);
+    }
+
+    /**
+     * Hook point to supply a (custom) TypeManager. By default a new TypeManager is created
+     */
+    @NotNull
+    private TypeManager createTypeManager() {
+        return new TypeManager();
+    }
+
+
+    /**
+     * Hook point to instantiate a {@link Monitor}. By default, the {@link ExtensionLoader#loadMonitor()} method is called.
+     * <p>
+     * <strong>Do not use this to provide a custom Monitor implementation! Use the extension mechanism for that! </strong>
+     */
+    @NotNull
+    private Monitor createMonitor() {
```

Comment:
load or create?

Reply:
I think `create` is fine, because the fact that the default runtime (i.e. `BaseRuntime`) actually loads a monitor based on all available `MonitorExtension` classes that are loaded using the `ServiceLoader` mechanism is an implementation detail. I'll update the comment.